Make crutest use BlockIO trait instead of a Guest

leftwo opened the pull request:
This enables future work where we plan to use a `Volume` in crutest instead of a `Guest`.

Update crutest to use `<T: BlockIO + Send + Sync + 'static>` instead of `Guest`.  
This involved adding a few more methods to the BlockIO trait, moving a few guest specific methods over to BlockIO, and adding support for these new methods in places where they did not have them.

This should result in no differences at run time as we are not doing anything different than before. 

I wanted this change first to reduce some of the boiler plate noise in future PRs.

mkeeter commented on crutest/src/main.rs:
```diff
@@ -804,6 +806,7 @@ async fn main() -> Result<()> {
         }
         _ => slog::Level::Info,
     };
+
```
nit: remove newline unless it's intentional, to minimize the diff

leftwo replied:
Fixed

mkeeter commented on crutest/src/main.rs:
```diff
@@ -844,7 +847,7 @@ async fn main() -> Result<()> {
     }
 
-    let _join_handle = up_main(crucible_opts, opt.gen, None, io, pr)?;
+    let _join_handle = up_main(crucible_opts.clone(), opt.gen, None, io, pr)?;
```
This seems to build fine without the added `.clone()`

leftwo replied:
Ah, that was leftover from when I did need it for code that I've since removed.
I'm suprosed clippy did not find that.

leftwo replied:
Fixed.

mkeeter commented:
Totally optional: It looks like we only need to pass in a `&Arc<T>` with `Send + Sync + 'static` constraints if we'll be using `tokio::spawn` in the `crutest` function, so we could take this opportunity to simplify a few `crutest` functions.

```diff
diff --git a/crutest/src/main.rs b/crutest/src/main.rs
index c32262df5e...ca5cd07518 100644
--- a/crutest/src/main.rs
+++ b/crutest/src/main.rs
@@ -931,7 +931,7 @@
     match opt.workload {
         Workload::Balloon => {
             println!("Run balloon test");
-            balloon_workload(&guest, &mut region_info).await?;
+            balloon_workload(guest.as_ref(), &mut region_info).await?;
         }
         Workload::Big => {
             println!("Run big test");
@@ -978,7 +978,7 @@
         Workload::Dirty => {
             println!("Run dirty test");
             let count = opt.count.unwrap_or(10);
-            dirty_workload(&guest, &mut region_info, count).await?;
+            dirty_workload(guest.as_ref(), &mut region_info, count).await?;
 
             /*
              * Saving state here when we have not waited for a flush
@@ -1680,8 +1680,8 @@
  * I named it balloon because each loop on a block "balloons" from the
  * minimum IO size to the largest possible IO size.
  */
-async fn balloon_workload<T: BlockIO + Send + Sync + 'static>(
-    guest: &Arc<T>,
+async fn balloon_workload<T: BlockIO>(
+    guest: &T,
     ri: &mut RegionInfo,
 ) -> Result<()> {
     for block_index in 0..ri.total_blocks {
@@ -1826,8 +1826,8 @@
  * Do a single random write to every extent, results in every extent being
  * touched without having to write to every block.
  */
-async fn fill_sparse_workload<T: BlockIO + Send + Sync + 'static>(
-    guest: &Arc<T>,
+async fn fill_sparse_workload<T: BlockIO>(
+    guest: &T,
     ri: &mut RegionInfo,
 ) -> Result<()> {
     let mut rng = rand_chacha::ChaCha8Rng::from_entropy();
@@ -2127,7 +2127,7 @@
     info!(log, "Begin replacement while reconciliation test");
     loop {
         info!(log, "[{c}] Touch every extent part 1");
-        fill_sparse_workload(guest, ri).await?;
+        fill_sparse_workload(guest.as_ref(), ri).await?;
 
         info!(log, "[{c}] Stop a downstairs");
         // Stop a downstairs, wait for dsc to confirm it is stopped.
@@ -2141,7 +2141,7 @@
             tokio::time::sleep(tokio::time::Duration::from_secs(4)).await;
         }
         info!(log, "[{c}] Touch every extent part 2");
-        fill_sparse_workload(guest, ri).await?;
+        fill_sparse_workload(guest.as_ref(), ri).await?;
 
         info!(log, "[{c}] Deactivate");
         guest.deactivate().await.unwrap();
@@ -2318,7 +2318,7 @@
     let mut new_ds = 3;
     for c in 1.. {
         info!(log, "[{c}] Touch every extent");
-        fill_sparse_workload(guest, ri).await?;
+        fill_sparse_workload(guest.as_ref(), ri).await?;
 
         guest.deactivate().await.unwrap();
         loop {
@@ -2466,7 +2466,7 @@
     assert!(targets.len() == 4);
 
     if fill {
-        fill_sparse_workload(guest, ri).await?;
+        fill_sparse_workload(guest.as_ref(), ri).await?;
     }
     // Make a copy of the stop at counter if one was provided so the
     // IO task and the replace task don't have to share wtq
@@ -2613,8 +2613,8 @@
  * We are trying to leave extents "dirty" so we want to exit before the
  * automatic flush can come through and sync our data.
  */
-async fn dirty_workload<T: BlockIO + Send + Sync + 'static>(
-    guest: &Arc<T>,
+async fn dirty_workload<T: BlockIO>(
+    guest: &T,
     ri: &mut RegionInfo,
     count: usize,
 ) -> Result<()> {
```

mkeeter commented on crutest/src/cli.rs:
```diff
@@ -223,8 +223,8 @@ enum CliCommand {
  */
-async fn cli_read(
-    guest: &Arc<Guest>,
+async fn cli_read<T: BlockIO + Send + Sync + 'static>(
```
See my comment in the main PR conversation; this only actually requires `BlockIO` and can take a `&T` instead of an `&Arc<T>`

leftwo replied:
Done

mkeeter commented on upstairs/src/lib.rs:
```diff
@@ -215,6 +219,17 @@ pub trait BlockIO: Sync {
         self.activate().await
     }
+    /// Conditional activate if not active.
+    async fn conditional_activate_with_gen(
```
It looks like this function is only needed by `Volume::activate_with_gen`, so I'd suggest removing it from the trait and just doing this logic in the body of that function (`volume.rs:615,626`).

mkeeter replied:
Same with `conditional_activate`, now that I look at it!

leftwo replied:
I've removed all the `conditional_activate()` and now the behavior is:
If we are active with the same gen as requested, then return OK()
If we are not active, then activate with the provide gen.
Else, return error.

mkeeter commented on upstairs/src/volume.rs:
```diff
@@ -610,6 +610,72 @@ impl BlockIO for Volume {
     }
+    async fn activate_with_gen(&self, gen: u64) -> Result<(), CrucibleError> {
+        for sub_volume in &self.sub_volumes {
+            sub_volume.conditional_activate_with_gen(gen).await?;
```
The correctness of this isn't obvious: why is it okay to consider yourself activated with a particular gen if a subvolume is activated with a _different_ gen?

leftwo replied:
So, `activate_with_gen()` has been updated to:
1) Allow activation with a higher number than the upstairs was started with, but only if the upstairs is not active yet.
2) If we are active, and the generation number we are active with is the same as what we requested, then return OK,
3) Any other combination should cause `activate_with_gen()` to fail.

I'd like to eventually remove `activate_with_gen()`, but not here in this PR till as that will involve some more serious test restructuring, and I'm not ready for that yet.

jmpesp commented on upstairs/src/volume.rs:
```diff
@@ -610,6 +610,72 @@ impl BlockIO for Volume {
+
+    // Return a vec of these?
+    // Return a struct with a vec for SV and Some/None for ROP?
+    async fn query_extent_size(&self) -> Result<Block, CrucibleError> {
```
What's this used for? I wonder why the caller is ask and if that could be changed.

leftwo replied:
We use it in crutest to construct specific kinds of tests where we cross the extent boundary with IOs.  It could be done other ways, like we pass the test itself information externally.  Having the test figure it out without having to supply anything is an easy way to do it.

leftwo replied:
The `sparse_fill()` is another good example.  We want to do an IO to every extent so they all end up dirty, but dont' want to have to do IO to every block in the extent.

leftwo commented on common/src/lib.rs:
```diff
@@ -77,7 +77,7 @@ pub enum CrucibleError {
     OffsetInvalid,
 
     #[error("Upstairs is already active")]
```
Done in 0a8efdd, thanks

leftwo commented on integration_tests/src/lib.rs:
```diff
@@ -3646,6 +3644,94 @@ mod test {
+
+    #[tokio::test]
+    async fn integration_test_guest_reactivate_same() -> Result<()> {
+        // Verify we fail activate_with_gen() if we are already active
```
done in 0a8efdd

leftwo commented:
I'll  fix the merge conflicts my merging main into the branch.
I did not want to do this till everyone had a chance to look at the changes so far.

jmpesp commented on upstairs/src/lib.rs:
```diff
@@ -104,11 +104,15 @@ pub const IO_OUTSTANDING_MAX_JOBS: usize = 10000;
 #[async_trait]
 pub trait BlockIO: Sync {
     async fn activate(&self) -> Result<(), CrucibleError>;
+    async fn activate_with_gen(&self, gen: u64) -> Result<(), CrucibleError>;
```
Eventually we want to get rid of `activate_with_gen` - for now, should we mark this as test-only, or hide it behind a flag? I'm also fine not doing this, so consider this a non-blocking suggestion :)

leftwo replied:
I want to focus on getting rid of it!

jmpesp commented on upstairs/src/pseudo_file.rs:
```diff
@@ -221,6 +210,16 @@ impl<T: BlockIO> CruciblePseudoFile<T> {
+        &self,
+        gen: u64,
+    ) -> Result<(), CrucibleError> {
+        self.block_io.activate_with_gen(gen).await
```
CruciblePseudoFile needs to do some queries to set some of its fields after activation, those need to be duplicated here.

leftwo commented:
from chat:
```
Alan (he/him)
You have any more thoughts on : https://github.com/oxidecomputer/crucible/pull/1452
I think I've addressed all comments.
James (he/him)
naw ship it
```
